### DISCUSSION TITLE
Get a v2.0 token for a service principal

### DISCUSSION THREAD

**Discussion by timothycrall** [Azure/ipam#305] [Q&A] [answered]
I'm able to connect to the IPAM API if I first log in to the UI and get a bearer token tied to my account.  But I need to automate connections from a script running within an ADO pipeline, and use a service principal instead of my personal account.

I've googled and tested a number of half-baked scripts confidently suggested by chatgpt, but everything I've tried is giving me a v1.0 token that the IPAM API rejects.  Can anyone provide a code example - using bash, powershell, python, whatever - of a way to get a v2.0 token for a service principal that the IPAM API will accept?

**Comment by DCMattyG** [comment, marked as answer]
Hey @timothycrall, just to verify did you manually create the App Registrations/Service Principals, or did you use the included automation scripts provided with this project?

The reason I ask is that the version of the token is set in the manifest of the App Registration and if not specifically set to "2" then it can default to v1. The scripting we use explicitly sets this to v2 upon creation of the App Registrations.

As far as generating a token, we have an example in our docs which can be found here:

https://azure.github.io/ipam/#/api/README?id=sample-api-calls

Additionally, please verify in the manifest of the App Registration the Token Version is set to "2" like so:

![image](https://github.com/user-attachments/assets/9b7d88eb-ce37-4cd0-850c-f6cb239db7a1)

Hope that helps, I'm always here to help further if required!

**Reply by timothycrall** [reply]
Thanks - that document seems like it has what I need. I didn't do the initial install so I'm not sure about the answer to a couple of your questions, but I was attempting to use a pre-existing service principal.  I'll check on the App Registration and see what's going on there.

**Reply by DCMattyG** [reply]
Great to hear, please let me know if you need anything else!
